pika/channel.py

```diff
@@ -1066,13 +1066,14 @@ def _on_close(self, method_frame):
         self._send_method(spec.Channel.CloseOk())
 
         if self.is_closing:
-            # Since we already sent Channel.Close, we need to wait for CloseOk
+            # Since we may have already sent Channel.Close, we need to wait for CloseOk
             # before cleaning up to avoid a race condition whereby our channel
             # number might get reused before our CloseOk arrives
 
             # Save the details to provide to user callback when CloseOk arrives
             self._closing_code_and_text = (method_frame.method.reply_code,
                                            method_frame.method.reply_text)
+            self._drain_blocked_methods_on_remote_close()
```

Comment:
Please add a code comment describing what prevents the blocked `Channel.Close` from becoming unblocked and sent to broker on an open connection so that it must be discarded this way manually.

Reply:
How about this comment?

https://github.com/pika/pika/blob/pika-945/pika/channel.py#L1286-L1289

Reply:
@lukebakken - that comment doesn't explain what is being asked.

Reply:
It sure doesn't, because it's the wrong set of lines! 😄 I meant to link to this:

https://github.com/pika/pika/blob/pika-945/pika/channel.py#L1292-L1295

I'll move the comment and make it clearer what the scenario is.

Reply:
Sorry, no -- it explains the side-effect, but not the root cause of the problem that we're working around with this code, which prevents reviewers/maintainers from gaining a valuable insight.

Here is what I consider to be the root cause of what this PR is fixing, which would be helpful to explain in a comment above the call to `self._drain_blocked_methods_on_remote_close`:

The server closes the channel before responding to an outstanding synchronous method (that possibly triggered the channel error and which isn't `Channel.Close`). AMQP 0.9.1 obliges the server to drop all methods arriving on a closed channel other than `Close-OK` and `Channel.Close`. Since the response to that synchronous method never arrives, the channel never becomes unblocked, and the `Channel.Close`, if any, in the blocked queue has no opportunity to be sent, and thus it's completion callback of the `Channel.Close` would never be called.